File: basic_scala/src/main/scala/pers/xiaoming/spark/core/MyAccumulator.scala
package pers.xiaoming.spark.core

import java.io.Serializable
import java.util.concurrent.atomic.AtomicLong

import org.apache.spark.util.AccumulatorV2

class MyAccumulator(val initValue:Long = 0) extends AccumulatorV2[Long, Long] with Serializable {

  private val myValue = new AtomicLong(initValue)

  override def isZero: Boolean = myValue.get() == 0

  override def copy(): AccumulatorV2[Long, Long] = new MyAccumulator(myValue.get())

  override def reset(): Unit = myValue.set(0)

  override def add(v: Long): Unit = myValue.addAndGet(v)

  override def merge(other: AccumulatorV2[Long, Long]): Unit = myValue.addAndGet(other.value)

  override def value: Long = myValue.get()
}

File: basic_java/src/main/java/pers/xiaoming/spark/core/MyAccumulator.java
package pers.xiaoming.spark.core;

import org.apache.spark.util.AccumulatorV2;

import java.io.Serializable;
import java.util.concurrent.atomic.AtomicLong;

class MyAccumulator extends AccumulatorV2<Long, Long> implements Serializable {
    private static final long serialVersionUID = 1L;

    private AtomicLong value;

    public MyAccumulator() {
        value = new AtomicLong(0);
    }

    public MyAccumulator(long value) {
        this.value = new AtomicLong(value);
    }

    @Override
    public boolean isZero() {
        return value.get() == 0;
    }

    @Override
    public AccumulatorV2<Long, Long> copy() {
        return new MyAccumulator(value.get());
    }

    @Override
    public void reset() {
        value.set(0L);
    }

    @Override
    public void add(Long v) {
        value.addAndGet(v);
    }

    @Override
    public void merge(AccumulatorV2<Long, Long> other) {
        value.addAndGet(other.value());
    }

    @Override
    public Long value() {
        return value.get();
    }
}
